Add account_tax_balance

eLBati opened the pull request:
See https://github.com/oca-sorrento/account-financial-reporting/blob/09bd2db8da2bda87870719d7456f449757b55488/account_tax_balance/README.rst

eLBati committed:
ADD account_tax_balance: Compute tax balances based on date range

IMP filters and views

Avoid create and delete taxes

ADD base balance

eLBati committed:
ADD README and images

FIX PEP8

FIX README image

gfcapalbo committed:
tests

eLBati committed:
FIX tests

FIX PEP8

use invoice._convert_to_write(invoice._cache). This way, the onchange will be inheritable and will add here also the added values

lepistone commented on account_tax_balance/tests/test_account_tax_balance.py:
```diff
@@ -0,0 +1,63 @@
+
+class TestAccountTaxBalance(TransactionCase):
+
+    def setUp(self):
```
not needed

lepistone commented on account_tax_balance/tests/test_account_tax_balance.py:
```diff
@@ -0,0 +1,63 @@
+            'invoice_line_tax_ids': [(6, 0, [tax.id])],
+        })
+        invoice._onchange_invoice_line_ids()
+        # save new() values to DB
```
?

eLBati replied:
@lepistone see `account.invoice._onchange_invoice_line_ids`
Do you know a better way to save to DB recordset created with `new()` ?

lepistone replied:
I don't know a better way. Then OK thanks!

pedrobaeza replied:
Please use `invoice._convert_to_write(invoice._cache)`. This way, the onchange will be inheritable and will add here also the added values. Example: https://github.com/OCA/commission/pull/67/files#diff-73f037396680f271cd1a6b014c9935ecR83

lepistone commented on account_tax_balance/tests/test_account_tax_balance.py:
```diff
@@ -0,0 +1,63 @@
+        }
+        invoice.tax_line_ids = None
+        self.env['account.invoice.tax'].create(tax_vals)
+        self.assertTrue((invoice.state == 'draft'))
```
`self.assertEqual('draft', invoice.state)`

lepistone commented on account_tax_balance/tests/test_account_tax_balance.py:
```diff
@@ -0,0 +1,63 @@
+        # change the state of invoice to open by clicking Validate button
+        invoice.signal_workflow('invoice_open')
+
+        self.assertEquals(tax.base_balance, -100)
```
for comparing floats assertAlmostEqual makes probably more sense.

eLBati replied:
In this case I want it to be -100

lepistone replied:
I understand, but normally with floats "equal" comparison is not relevant. It might work here with =, especially because integers can be represented exactly, but it is not always the case. That is because a base-10 decimal number like 0.5 is represented in a float only by approximation IFRC.

I am suggesting to use `assertAlmostEqual` with the default setting (which is something like 7 decimal places) just to prevent low-level problems with floats, not with something like 2 decimal places because that could hide real bugs.

Thanks!

eLBati replied:
I understand `assertAlmostEqual` is useful when approximation risks are present but, in this case, I can't see those risks. (What does IFRC stand for?)

lepistone replied:
If I Remember Correctly.

As I said, floats are represented exactly only in few cases (including integers) otherwise they are approximations. "Approximation risks" are almost everywhere and an exact comparison of floats doesn't make sense in general.

I can live with that tough since `-100` is an integer.

Thanks!

lepistone commented on account_tax_balance/wizard/open_tax_balances_view.xml:
```diff
@@ -0,0 +1,40 @@
+            parent="account.menu_finance_reports"
+            groups="account.group_account_user,account.group_account_manager"></menuitem>
+    </data>
+</openerp>
```
newline

lepistone commented:
minor remarks, otherwise looks good, thanks! 👍

jgrandguillaume commented:
Hi,

Thanks, functional tests ok on my side.

:+1: 

Regards,

Joël

lepistone commented on account_tax_balance/models/account_tax.py:
```diff
@@ -0,0 +1,86 @@
+            company_id = self.env.user.company_id.id
+        else:
+            company_id = self.env.context['company_id']
+        return from_date, to_date, company_id, target_move
```
What do you think about:

``` python
context = self.env.context
return (
    context.get('from_date') or fields.Date.context_today(self),
    context.get('to_date') or fields.Date.context_today(self),
    context.get('target_move') or 'posted',
    context.get('company_id') or self.env.user.company_id.id,
)
```

pedrobaeza replied:
It's even easier:

```
context = self.env.context
return (
    context.get('from_date', fields.Date.context_today(self)),
    context.get('to_date', fields.Date.context_today(self)),
    context.get('target_move', 'posted'),
    context.get('company_id', self.env.user.company_id.id),
)
```

and you also allow False values in the context.

lepistone commented on account_tax_balance/models/account_tax.py:
```diff
@@ -0,0 +1,86 @@
+
+    def _compute_balance(self):
+        from_date, to_date, company_id, target_move = self.get_context_values()
+        for tax in self:
```
By splitting the two methods `compute_balance` and `compute_base_balance` you introduced duplication. If you did everything together you could just write the `balance` and `base_balance` in-place in the same method instead of returning the value and writing it here. What do you think?

lepistone commented:
Thanks for your work @eLBati @gfcapalbo ! please give a look to the suggestions (which are minor). Then I'd like to merge this.

eLBati committed:
better get_context_values

unify method compute_balance

ADD open move lines linked to balance

eLBati commented:
@lepistone @pedrobaeza thanks for your suggestions, I made the changes

pedrobaeza commented on account_tax_balance/__openerp__.py:
```diff
@@ -0,0 +1,25 @@
+    "version": "9.0.1.0.0",
+    "category": "Accounting & Finance",
+    "website": "https://www.agilebg.com/",
+    "author": "Agile Business Group, Odoo Community Association (OCA)",
```
Therp is not credited here?

pedrobaeza commented on account_tax_balance/README.rst:
```diff
@@ -0,0 +1,60 @@
+Usage
+=====
+
+Accounting --> Reporting --> Open Tax Balances
```
Why the "Open" word? Isn't better to just be "Taxes balance"?

pedrobaeza commented:
For improving the functionality, you should save in a many2many the reference to all the account.move.line that compose the balance and the base balance, and provide buttons at line level that allows to open these lines for tracing the source of the data.

lepistone commented:
@pedrobaeza I'd like this to be merged quickly if it works, is it OK for you if we merge without the last thing you suggest? Thanks a lot!

pedrobaeza commented:
That functionality is basic, because the report is not useful without that, so I consider it a blocking point.

fclementic2c commented:
First thank you for this very important module existing in enterprise version but not in community version.

My comments : 
- I notices a sign issue on this view : it should be - Debit + Credit (and not Debit - Credit as usual in accounting journal entries) since you want to see purchases as negative value for ex. Otherwise total VAT to be paid is wrong !
- I totally agree with Predro's comment. Without this, this module is not helpful. You must be able to dig into lines to find where are errors !

thanks

eLBati committed:
IMP terms and author

eLBati commented:
@pedrobaeza thanks, I made the changes.
I added 2 buttons ('View tax lines' and 'View base lines') for every tax.

@fclementic2c I just used the `balance` field of `account.move.line`
see https://github.com/oca-sorrento/account-financial-reporting/blob/dcc68e6932fd2cee8f4cae94417a10bc741a0e93/account_tax_balance/models/account_tax.py#L48
Should I use its opposite?

Thanks

fclementic2c commented:
@eLBati Yes, on the VAT return you should use the opposite of the balance field.

balance is debit - credit whereas on tax return you want to see what vat has to be paid so : VAT on sales (credit) - VAT on purchases (debit).

eLBati committed:
FIX
balance is debit - credit whereas on tax return you want to see what vat has to be paid so : VAT on sales (credit) - VAT on purchases (debit).

eLBati commented:
@fclementic2c thanks, done

eLBati committed:
ADD account_tax_balance: testing wizard with date range

ADD account_tax_balance: testing buttons

IMP account_tax_balance: covering specific methods with tests

archetipo commented on account_tax_balance/views/account_tax_view.xml:
```diff
@@ -0,0 +1,49 @@
+            <field name="search_view_id" ref="view_tax_search_balance"/>
+        </record>
+    </data>
+</openerp>
```
no new line

eLBati replied:
@archetipo thanks, done

archetipo reviewed (approved):
@eLBati thanks, only one little remark

antespi commented:
In the Spanish localization (and in other ones I guess) we need to separate taxes from normal sale/purchase and refunds (sale or purchase), so we want to propose these changes:
- Make 4 taxes columns:
  - Tax base
  - Tax amount
  - Tax refund base
  - Tax refund amount

In the other hand, It would be great to show only lines with amount in any column, because in some localization there are many taxes

We can do these changes, if you don't mind, and PR them to 'oca-sorrento' repo

gurneyalex reviewed (approved):
code review only, no tests done.

antespi committed:
Consider normal and refund operations separately

Allow to explore all move lines

Spanish translation

Bugfixes

Show negative lines too

Show move type in account.move views

Tests include new fields

antespi commented:
@eLBati I've already done our propose, please review it in oca-sorrento repo: https://github.com/oca-sorrento/account-financial-reporting/pull/11

flotho commented:
Hi, I've tested it on runbot successfully.
Regards

eLBati commented:
Merged https://github.com/oca-sorrento/account-financial-reporting/pull/11
Thanks @antespi @pedrobaeza

pedrobaeza commented:
One last task before merging, please: you know my preference for squashing commits grouped by author and logic :wink:

antespi commented:
@eLBati For mi is ok, you can squash my commits in only one

eLBati commented:
@pedrobaeza  squashed!

pedrobaeza commented:
Thanks!
